contracts/src/StakeTable.sol

```diff
@@ -52,493 +102,356 @@ contract StakeTable is AbstractStakeTable, Ownable, InitializedAt {
+        // Mark funds as spent
+        delete undelegations[validator][msg.sender];
 
-        emit Exit(msg.sender, exitEpoch);
-    }
+        SafeTransferLib.safeTransfer(token, msg.sender, amount);
 
-    /// @notice Withdraw from the staking pool. Transfers occur! Only successfully exited keys can
-    /// withdraw past their `exitEpoch`. Validators have to first call requestExit to be assigned an
-    /// exit epoch
-    /// @return The total amount withdrawn, equal to `Node.balance` associated with `blsVK`
-    /// TODO: add epoch logic so that we can ensure the node has first requested to exit and waiting
-    /// for the exit escrow period to be over
-    function withdrawFunds() external virtual override returns (uint256) {
-        Node memory node = nodes[msg.sender];
-
-        if (node.account == address(0)) {
-            revert NodeNotRegistered();
-        }
+        emit Withdrawal(msg.sender, amount);
+    }
 
-        if (node.account != msg.sender) {
-            revert Unauthenticated();
+    /// @notice Withdraw previously delegated funds after a validator has exited
+    /// @param validator The validator to withdraw from
+    function claimValidatorExit(address validator) external virtual {
+        uint256 unlocksAt = validatorExits[msg.sender];
+        if (unlocksAt == 0) {
+            revert ValidatorNotExited();
         }
 
-        uint256 balance = node.balance;
-        if (balance == 0) {
-            // then there's nothing to withdraw but revert so that they're aware that the withdrawal
-            // failed
-            revert InsufficientStakeBalance(0);
+        if (block.timestamp < unlocksAt) {
+            revert PrematureWithdrawal();
         }
 
-        // Verify that the exit escrow period is over.
-        if (currentEpoch() < node.exitEpoch + exitEscrowPeriod(node)) {
-            revert PrematureWithdrawal();
+        uint256 amount = delegations[validator][msg.sender];
+        if (amount == 0) {
+            revert NothingToWithdraw();
         }
-        totalStake -= balance;
 
-        // Delete the node from the stake table.
-        delete nodes[msg.sender];
+        // Mark funds as spent
+        delegations[validator][msg.sender] = 0;
 
-        // Transfer the balance to the node's account.
-        SafeTransferLib.safeTransfer(ERC20(tokenAddress), node.account, balance);
+        SafeTransferLib.safeTransfer(token, msg.sender, amount);
 
-        return balance;
+        emit Withdrawal(msg.sender, amount);
     }
 
     /// @notice Update the consensus keys for a validator
     /// @dev This function is used to update the consensus keys for a validator
-    /// @dev This function can only be called by the validator itself when it's not in the exit
-    /// queue
+    /// @dev This function can only be called by the validator itself when it hasn't exited
+    ///      TODO: MA: is this a good idea? Why should key rotation be blocked for an exiting
```

Comment:
hmm, why would a validator rotate their keys if they are exiting and no longer intend to vote in consensus - unless they lost their keys and that's why they're exiting? but then it's either you exit or you update your keys? any thoughts on this @rob-maron

Reply:
In was thinking in theory they could discover a key leak after registering and exit and want to rotate their keys. However I realize now that wouldn't take effect until they exited the stake table anyway so that key change would be useless.